posix spawn, take 2

vtjnash opened the pull request:
This is a commit series to re-enable the usage of posix_spawn on macOS.

CI: https://ci.nodejs.org/job/libuv-test-commit-osx-cmake/359/
CI+node: https://ci.nodejs.org/job/libuv-in-node/191/

vtjnash commented on src/unix/process.c:
```diff
@@ -342,6 +354,447 @@ static void uv__process_child_init(const uv_process_options_t* options,
+  }
+
+  if (seen_eacces)
+    return EACCESS;
```
```suggestion
    return EACCES;
```

vtjnash commented on src/unix/process.c:
```diff
@@ -342,6 +354,447 @@ static void uv__process_child_init(const uv_process_options_t* options,
+  int seen_eacces;
+
+  path = NULL;
+  err = -1;
```
```suggestion
  err = -1;
  seen_eacces = 0;
```

vtjnash commented:
darn, of course we would now run into what looks like another pre-existing nodejs/kernel bug more frequently now 😂

deepak1556 commented:
@vtjnash there is also an issue with `posix_spawnp` code path when used with `posix_spawn_file_actions_addchdir_np` which turns out to be a bug in the implementation. Originally caught in https://github.com/microsoft/vscode/issues/118267#issuecomment-797016213. I am using `pthread_chdir_np` as a workaround.

```patch
diff --git a/src/unix/process.c b/src/unix/process.c
index 0fb9fda1..ef290d0e 100644
--- a/src/unix/process.c
+++ b/src/unix/process.c
@@ -366,6 +366,28 @@ static uv_once_t posix_spawn_init_once = UV_ONCE_INIT;
 static uv__posix_spawn_fncs_t posix_spawn_fncs;
 static int posix_spawn_can_use_setsid;
 
+static int uv__change_current_thread_directory(const char* path) {
+  // Changes the current thread's directory to a path or directory file
+  // descriptor. libpthread only exposes a syscall wrapper starting in
+  // macOS 10.12, but the system call dates back to macOS 10.5. On older OSes,
+  // the syscall is issued directly.
+  if (__builtin_available(macOS 10.12, *)) {
+    return pthread_chdir_np(path);
+  } else {
+    return syscall(SYS___pthread_chdir, path);
+  }
+}
+
+// The recommended way to unset a per-thread cwd is to set a new value to an
+// invalid file descriptor, per libpthread-218.1.3/private/private.h.
+static int uv__reset_current_thread_directory() {
+  if (__builtin_available(macOS 10.12, *)) {
+    return pthread_fchdir_np(-1);
+  } else {
+    return syscall(SYS___pthread_fchdir, -1);
+  }
+}
+
 
 void uv__spawn_init_posix_spawn_fncs(void) {
   /* Try to locate all non-portable functions at runtime */
@@ -501,12 +523,23 @@ int uv__spawn_set_posix_spawn_file_actions(posix_spawn_file_actions_t* actions,
 
   /* Set the current working directory if requested */
   if (options->cwd != NULL) {
-    if (posix_spawn_fncs->file_actions.addchdir_np == NULL) {
-      err = ENOSYS;
-      goto error;
-    }
+    if (strchr(options->file, '/') != NULL) {
+      /* There is a bug in macOS where a relative executable
+       * path like "../myprogram" will cause `posix_spawnp` to
+       * successfully launch the program, but erroneously return
+       * ENOENT when used with posix_spawn_file_actions_addchdir_np
+       * because of the stat call performed on the executable path
+       * which was introduced in macOS 10.15.
+       * Ref: https://opensource.apple.com/source/Libc/Libc-1439.40.11/sys/posix_spawn.c.auto.html */
+      err = uv__change_current_thread_directory(options->cwd);
+    } else {
+      if (posix_spawn_fncs->file_actions.addchdir_np == NULL) {
+        err = ENOSYS;
+        goto error;
+      }
 
-    err = posix_spawn_fncs->file_actions.addchdir_np(actions, options->cwd);
+      err = posix_spawn_fncs->file_actions.addchdir_np(actions, options->cwd);
+    }
     if (err != 0)
       goto error;
   }
@@ -628,6 +661,9 @@ int uv__spawn_resolve_and_spawn(const uv_process_options_t* options,
     do
       err = posix_spawnp(pid, options->file, actions, attrs, options->args, env);
     while (err == EINTR);
+    /* Restore the thread's working directory if it was changed. */
+    if (options->cwd != NULL)
+      uv__reset_current_thread_directory();
     return err;
   }
 
```

vtjnash commented:
The bug here sounds almost identical to this reported problem from 2006: https://gcc.gnu.org/legacy-ml/gcc/2006-09/msg00631.html. Is there anyone here who can tell me if rdar://problem/4736203 is still a known issue in the mach kernel? I can provide code to reproduce, but currently involves re-building all of nodejs with this PR and then running one particular unit test a small number of times.

vtjnash commented:
I can reproduce this kernel bug without libuv also simply by calling posix_spawn a few times:
https://gist.github.com/vtjnash/679174286cf282c298e5d11752343b8a

The specific conditions under which it occurs seems to be if posix_spawn returns an error, it will revoke any pending SIGCHLD, so we might be able to work around it by calling `if (posix_spawn() == 0) kill(getpid(), SIGCHLD);`, but it is awful and incomplete, and really seems to require an Apple kernel developer to correct this.

jpcanepa commented:
I hope this time the code lands. Ping me if I can help somehow, I may manage to squeeze some time.

deepak1556 commented:
Adding here for visibility, FB9578526 tracks https://github.com/libuv/libuv/pull/3257#issuecomment-891374523

vtjnash commented on src/unix/process.c:
```diff
@@ -423,48 +877,45 @@ int uv_spawn(uv_loop_t* loop,
-    abort();
-  }
+  /* Spawn the child */
+  exec_errorno = uv__spawn_and_init_child(options, stdio_count, pipes, signal_pipe[1], &pid);
```
Review comment from Apple: `signal_pipe` is only needed on the `fork` code path, and can be moved there as it is not needed by the posix_spawn API.

KishanBagaria commented:
What's blocking this – how can we help?

KishanBagaria commented:
https://github.com/libuv/libuv/pull/3064#issuecomment-778495619 says "we had to revert this because it was causing failures in the Node.js test suite."

Can we use `posix_spawn` when it's working fine (not triggering the bug / failure in test suite) and the original function in all other cases as a stopgap? This should be what most people want until we have the proper fix.

vtjnash commented:
It is a show-stopper-type bug that breaks all usages.

1bitphoenix commented:
@vtjnash 
Can you please provide some context/idea around what that old kernel bug in xnu bug actually is, if it isn't resolved by the patch [#3257 (comment)](https://github.com/libuv/libuv/pull/3257#:~:text=visibility%2C%20FB9578526%20tracks-,%233257%20(comment),-deepak1556%20mentioned%20this)?

vtjnash commented:
mach will sometimes fail to deliver signals as required, and as needed for proper function of `uv_spawn`. This seems to be greatly exacerbated by using posix_spawn. A partial workaround would be to switch to kevent notifications of process exit, but that is possibly a non-trivial amount of work, and only reduces the scope of the issue, but doesn't solve it.

crrodriguez commented:
Just FYI: This patch is also the best path forward on linux-glibc, not just on macos, however further changes are needed. namely:

POSIX_SPAWN_CLOEXEC_DEFAULT --> is not implemented, however the same practical effect is calling posix_spawn_file_actions_addclosefrom_np closing fd =>3  if that function is available,

using posix spawn on recent glibc's will give you the best performing way while using significantly less memory, it is also not affected  by all the vfork() problems because it is implemented using clone CLONE_VFORK  with custom stack..

nornagon commented:
It looks like Chromium uses posix_spawn on macOS to launch processes: https://source.chromium.org/chromium/chromium/src/+/main:base/process/launch_mac.cc;l=295;drc=7319bbdb7d6ea97ef3237381e4a88b7a465dd0da

Is there something that libuv needs that Chromium's code doesn't address? Can we take inspiration from there?

vtjnash commented:
As long as you don't care what happens if the process fails to launch, it works fine already

vtjnash commented:
Chromium uses EVFILT_PROC (instead of SIGCHLD), which also solves the issue. We could use that too, if someone was willing to put in the work to implement it.

nornagon commented:
> Chromium uses EVFILT_PROC (instead of SIGCHLD), which also solves the issue. We could use that too, if someone was willing to put in the work to implement it.

I have some time to spend on this, but I don't have a lot of background context here. Would you be able to give me some pointers on what would need to change to use EVFILT_PROC as opposed to SIGCHLD?

vtjnash commented:
We need to change the way libuv registers for the NOTE_EXITSTATUS for these objects: by calling EV_ADD here for them, and then handling them directly when kqueue notifies us of them, instead of waiting for the SIGCHLD event and calling `waitpid` on them (and doing it for *BSD's only, including macOS, but not changing linux or others).

nornagon commented:
@vtjnash Gotcha, thanks. How have you been reproducing the failure? The test suite seems to pass reliably on my machine.

vtjnash commented:
The code is above in a gist https://gist.github.com/vtjnash/679174286cf282c298e5d11752343b8a

nornagon commented:
@vtjnash sorry, I meant for triggering it in libuv's usage 😅  I'd like to be able to run some sort of test on libuv to check (a) that the current usage in this patch triggers the bug and (b) whether the code I end up writing has, in fact, fixed the issue. It looks like the gist you posted is a standalone example of triggering the bug.

vtjnash commented:
Oh, I do not have that version written. I have only tested with the nodejs test and that snippet.

nornagon commented:
Ah, the nodejs test suite triggers it? Okay, that’s a good lead, I can
hopefully distill a test for libuv from that.

On Thu, Jan 20, 2022 at 17:21 Jameson Nash ***@***.***> wrote:

> Oh, I do not have that version written. I have only tested with the nodejs
> test and that snippet.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/libuv/libuv/pull/3257#issuecomment-1018075718>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AABKGAB5SRE4E27V3FW47VLUXCYK5ANCNFSM5BAUTZUQ>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
> You are receiving this because you commented.Message ID:
> ***@***.***>
>
-- 

j

nornagon commented:
Okay, I've narrowed it down to the `parallel/test-child-process-exec-error` test from the Node.js test suite. Without this patch applied, that test passes consistently (albeit slowly). With the patch applied, it _sometimes_ hangs, i.e. doesn't exit. Seems to be fairly unreliable, fails maybe 1% of the time.

Next step is to extract this test into a test in libuv, and hopefully produce a more reliably-failing version.

nornagon commented:
Here's a libuv test that exhibits the behavior:
```patch
diff --git a/test/test-spawn.c b/test/test-spawn.c
index 9f2eb24b..50022bce 100644
--- a/test/test-spawn.c
+++ b/test/test-spawn.c
@@ -1891,6 +1891,43 @@ TEST_IMPL(spawn_quoted_path) {
 #endif
 }
 
+TEST_IMPL(spawn_exercise_sigchld_issue) {
+  int r;
+
+  init_process_options("spawn_helper1", exit_cb);
+
+  r = uv_spawn(uv_default_loop(), &process, &options);
+  ASSERT(r == 0);
+
+  // This test exercises a bug in the darwin kernel that causes SIGCHLD not to
+  // be delivered sometimes. Calling posix_spawn many times increases the
+  // likelihood of encountering this issue, so spin a few times to make this
+  // test more reliable.
+  for (int i = 0; i < 100; i++) {
+    uv_process_options_t options = {0};
+    uv_process_t process;
+    options.file = args[0] = "program-that-had-better-not-exist";
+    args[1] = NULL;
+    args[2] = NULL;
+    args[3] = NULL;
+    args[4] = NULL;
+    options.args = args;
+    options.exit_cb = fail_cb;
+    options.flags = 0;
+    r = uv_spawn(uv_default_loop(), &process, &options);
+    ASSERT(r == UV_ENOENT || r == UV_EACCES);
+  }
+
+  r = uv_run(uv_default_loop(), UV_RUN_DEFAULT);
+  ASSERT(r == 0);
+
+  ASSERT(exit_cb_called == 1);
+  ASSERT(close_cb_called == 1);
+
+  MAKE_VALGRIND_HAPPY();
+  return 0;
+}
+
 /* Helper for child process of spawn_inherit_streams */
 #ifndef _WIN32
 void spawn_stdin_stdout(void) {
```

nornagon commented:
I've opened #3441 as a continuation of this PR, with the new EVFILT_PROC / NOTE_EXIT code. It's a bit hacky as I'm not all that familiar with the libuv codebase, but it at least passes my new test :)

vtjnash commented:
CI: https://ci.nodejs.org/job/libuv-test-commit/2219/

I have rebased, addressed Apple's review comment, and fixed an unrelated resource management issue (with zombies) on an unlikely error path. Should be good to merge now, if we are passing all CI.

vtjnash commented:
libuv CI is green, so let's check nodejs: https://ci.nodejs.org/job/libuv-in-node/199/

vtjnash commented:
Ah, that did not actually go well. I have created https://github.com/libuv/libuv/pull/3451 to fix that failure (a small problem with https://github.com/libuv/libuv/pull/3441 @nornagon can you review it?)

vtjnash commented:
Trying this again:
CI+node: https://ci.nodejs.org/job/libuv-in-node/200/
CI: https://ci.nodejs.org/job/libuv-test-commit-freebsd/2105/
CI: https://ci.nodejs.org/job/libuv-test-commit-osx-cmake/397/

vtjnash commented:
Reopenning, since GitHub internal state was confused. The only failures above look like unrelated CI problems with nodejs and libuv. This is ready to merge, but would be good to get a review first of the new commits (relative to #3064).

santigimeno commented on src/unix/process.c:
```diff
@@ -348,25 +361,551 @@ static void uv__process_child_init(const uv_process_options_t* options,
+      goto error;
+  }
+
+  /* First, duplicate any required fd into orbit, out of the range of
+   * the descriptors that should be mapped in. */
+  for (fd = 0; fd < stdio_count; fd++) {
+    if (pipes[fd][1] < 0)
+      continue;
+
+    err = posix_spawn_file_actions_adddup2(
+      actions,
+      pipes[fd][1],
+      stdio_count + fd);
+    if (err != 0)
+      goto error;
+  }
```
A couple of questions here as I'm not familiar with this api:
- Is there a reason we're duplicating every fd instead of the lower fd's only as it's done [here](https://github.com/libuv/libuv/blob/v1.x/src/unix/process.c#L260-L265).
- Is there a chance that when duplicating the fd's into `stdio_count+fd` we might be closing a currently opened fd with that value?

vtjnash replied:
Good point. I think the original hope here was that we were moving the files in sorted order, but that does not appear to be ensured anywhere in the code, so this seems to need to be fixed.

santigimeno reviewed (approved):
Couldn't test it but code looks good to me, just left some comments / questions

vtjnash commented on src/unix/process.c:
```diff
@@ -220,10 +233,11 @@ static void uv__write_errno(int error_fd) {
-                                   int stdio_count,
-                                   int (*pipes)[2],
-                                   int error_fd) {
+static void uv__process_child_init(
+        const uv_process_options_t* options,
+        int stdio_count,
+        int (*pipes)[2],
+        int error_fd) {
   sigset_t signewset;
```
The lines were getting too long for the 80 character limit. I don't actually remember though how much indentation to use for the subsequent lines though, so it defaulted to 1 tab width, though I can fix that.

vtjnash commented on src/unix/process.c:
```diff
@@ -348,25 +361,551 @@ static void uv__process_child_init(const uv_process_options_t* options,
+  /* Parse the version major as a number. If it is greater than
+   * the major version for macOS Catalina (aka macOS 10.15), then
+   * the POSIX_SPAWN_SETSID flag is available */
+  version_major = atoi(version_major_str);
```
IIUC, those are defined to be the same function. Though you made me realize we might want to specify the locale to this function:

```suggestion
  version_major = atoi_l(version_major_str, LC_C_LOCALE);
```

vtjnash commented on src/unix/process.c:
```diff
@@ -348,25 +361,551 @@ static void uv__process_child_init(const uv_process_options_t* options,
+  if (err != 0)
+    goto error;
+
+  err = uv__spawn_set_posix_spawn_file_actions(
+    &actions,
+    posix_spawn_fncs,
+    options,
+    stdio_count,
+    pipes);
```
The `err` variable was longer originally (`exec_errorno`), which meant it could not fit that style without going over 80 characters on the line. It would fit now.

vtjnash commented on src/unix/process.c:
```diff
@@ -348,25 +361,551 @@ static void uv__process_child_init(const uv_process_options_t* options,
+  int r;
+  int version_major;
+
+  /* By default, assume failure */
+  posix_spawn_can_use_setsid = 0;
```
agreed, this should be removed

vtjnash commented on src/unix/process.c:
```diff
@@ -348,25 +361,551 @@ static void uv__process_child_init(const uv_process_options_t* options,
+  static const int MACOS_CATALINA_VERSION_MAJOR = 19;
+  char version_str[256];
+  char* version_major_str;
+  size_t version_str_size = 256;
```
We never separated declaration from assignment (some things may be impossible to write then), but the style guide does require separation of code and declarations. Constants (and other C-style constexprs) are permitted everywhere.

vtjnash commented:
@nornagon if you have more time and interest in helping with this, the last remaining blocker now to merging this is to alter the posix_spawn_file_actions_adddup2 to be more careful to pick a dup2 target fd which does not conflict with an existing fd that is needed by another action. (https://github.com/libuv/libuv/pull/3257#discussion_r797553011)

nornagon commented:
@vtjnash I'm happy to, but would you be able to help with a little more context? e.g. would it be sufficient to sort the `pipes` list somehow?

vtjnash commented:
There is not a sort order that will work (for example, the mapping 0=>0, 2=>1, 1=>2, 2=>3). The code can look similar to the non-posix_spawn case, and first needs to dup2 any low numbered fd (less than stdio_count) to a free fd number, then to the final position. Most of that code even is already written, but it fails to check that the number is actually free, which it can do by scanning the input array to check for any conflicts. Relatedly, it can also avoid calling dup2 on any fd > stdio_count, which may help simplify this code. It is always a very small array, so any n^2 behavior is fine here.

vtjnash commented:
@santigimeno what do you think now? I think I have improved the code to also avoid a couple redundant syscalls (`close` is not needed because CLOEXEC is set).

santigimeno commented on src/unix/process.c:
```diff
@@ -275,22 +275,22 @@ static void uv__process_child_init(
+        /* Redirect stdin, stdout and stderr to /dev/null even if UV_IGNORE is
+         * set. */
+        uv__close_nocheckstdio(fd); /* Free up fd, if it happens to be open. */
         use_fd = open("/dev/null", fd == 0 ? O_RDONLY : O_RDWR);
```
Do you think `O_CLOEXEC` could also be used here?

vtjnash replied:
yeah, I agree it could, but didn't seem beneficial, and there's a slight chance some kernel won't support it

santigimeno commented on src/unix/process.c:
```diff
@@ -220,10 +234,11 @@ static void uv__write_errno(int error_fd) {
-                                   int stdio_count,
-                                   int (*pipes)[2],
-                                   int error_fd) {
+static void uv__process_child_init(
+    const uv_process_options_t* options,
+    int stdio_count,
+    int (*pipes)[2],
+    int error_fd) {
```
Still I don't understand the style code change here. It was ok before and if I see things correctly the code hasn't really changed.

vtjnash replied:
I wanted to keep consistency within the file, since several of the new calls here now must be wrapped this way to fit the line-width constraint. This is how it is specified to wrap them in the libuv style guide: https://google.github.io/styleguide/cppguide.html#Function_Declarations_and_Definitions

santigimeno reviewed (approved):
LGTM with a question and a nit.

vtjnash commented on src/unix/process.c:
```diff
@@ -348,25 +358,575 @@ static void uv__process_child_init(const uv_process_options_t* options,
+  /* Parse the version major as a number. If it is greater than
+   * the major version for macOS Catalina (aka macOS 10.15), then
+   * the POSIX_SPAWN_SETSID flag is available */
+  version_major = atoi_l(version_major_str, LC_C_LOCALE);
```
From CI:
```
/Users/iojs/build/workspace/libuv-test-commit-osx-cmake/nodes/osx1014/src/unix/process.c:403:45: error: use of undeclared identifier 'LC_C_LOCALE'
  version_major = atoi_l(version_major_str, LC_C_LOCALE);
                                            ^
1 error generated.
make[2]: *** [CMakeFiles/uv.dir/src/unix/process.c.o] Error 1
```

This is just a name for NULL:
`#define LC_C_LOCALE				((locale_t)NULL)`

As specified in the man page:

> If a NULL locale_t is passed, the C locale will be used.

So we can use that value directly, rather than the convenience name add in 10.15
.
```suggestion
  version_major = atoi_l(version_major_str, NULL); /* Use LC_C_LOCALE */
```

vtjnash commented on src/unix/process.c:
```diff
@@ -220,10 +234,11 @@ static void uv__write_errno(int error_fd) {
-                                   int stdio_count,
-                                   int (*pipes)[2],
-                                   int error_fd) {
+static void uv__process_child_init(
+    const uv_process_options_t* options,
+    int stdio_count,
+    int (*pipes)[2],
+    int error_fd) {
```
```suggestion
static void uv__process_child_init(const uv_process_options_t* options,
                                   int stdio_count,
                                   int (*pipes)[2],
                                   int error_fd) {
```

vtjnash commented on src/unix/process.c:
```diff
@@ -348,25 +358,575 @@ static void uv__process_child_init(const uv_process_options_t* options,
+}
+
+
+static int uv__spawn_resolve_and_spawn(
+    const uv_process_options_t* options,
+    posix_spawnattr_t* attrs,
+    posix_spawn_file_actions_t* actions,
+    pid_t* pid) {
```
```suggestion
static int uv__spawn_resolve_and_spawn(const uv_process_options_t* options,
                                       posix_spawnattr_t* attrs,
                                       posix_spawn_file_actions_t* actions,
                                       pid_t* pid) {
```

vtjnash commented on src/unix/process.c:
```diff
@@ -348,25 +358,575 @@ static void uv__process_child_init(const uv_process_options_t* options,
+      }
+    }
+
+    err = posix_spawn_file_actions_adddup2(actions, use_fd, fd);
```
`posix_spawn_file_actions_adddup2` is undefined behavior if use_fd == fd. This is what broke the nodejs build.

```suggestion
    if (fd == use_fd)
        err = posix_spawn_file_actions_addinherit_np(actions, fd);
    else
       err = posix_spawn_file_actions_adddup2(actions, use_fd, fd);
```

The question is whether we want to deal with it using with the nonportable function, or change the previous loop to copy them up then back.

santigimeno replied:
Isn't this part of the code already only defined for `__APPLE__` or am I missing something?

vtjnash replied:
I am hoping to enable this for all other platforms too (after merging), but let's go with the np extension for now then

KishanBagaria commented:
Really eager to see this merged (and not have to maintain an Electron fork) anymore. Huge thanks to everyone contributing to this! 😄

vtjnash commented:
https://ci.nodejs.org/job/libuv-in-node/208/
https://ci.nodejs.org/job/libuv-test-commit-osx-cmake/407/

richardlau commented:
This has broken AIX, IBM i and z/OS CI (fyi @libuv/aix @libuv/ibmi @libuv/zos),
e.g. https://ci.nodejs.org/view/libuv/job/libuv-test-commit/2228/

https://ci.nodejs.org/job/libuv-test-commit-ibmi-cmake/935/BUILDTYPE=Release,nodes=ibmi73-ppc64/console
```console
/home/IOJS/build/workspace/libuv-test-commit-ibmi-cmake/BUILDTYPE/Release/nodes/ibmi73-ppc64/src/unix/process.c: In function 'uv__process_child_init':
/home/IOJS/build/workspace/libuv-test-commit-ibmi-cmake/BUILDTYPE/Release/nodes/ibmi73-ppc64/src/unix/process.c:278:34: error: 'F_DUPFD_CLOEXEC' undeclared (first use in this function)
  278 |     pipes[fd][1] = fcntl(use_fd, F_DUPFD_CLOEXEC, stdio_count);
      |                                  ^~~~~~~~~~~~~~~
/home/IOJS/build/workspace/libuv-test-commit-ibmi-cmake/BUILDTYPE/Release/nodes/ibmi73-ppc64/src/unix/process.c:278:34: note: each undeclared identifier is reported only once for each function it appears in
```

vtjnash commented:
According to the latest documentation, that has been a part of POSIX since 2001

bnoordhuis commented:
F_DUPFD_CLOEXEC is defined in POSIX.2008 but I can see why you'd think it's from the previous edition: it was approved in Austin Group Interpretation 1003.1-2001 # 171 from 2007.

@richardlau What might work is compiling with `-D_POSIX_C_SOURCE=200809`, maybe `-D_XOPEN_SOURCE=600` or `-D_XOPEN_SOURCE=700` too.

richardlau commented:
The situation on AIX is more subtle -- the builds on AIX 7.1 are failing, but on 7.2 they are passing. I found https://www.ibm.com/support/pages/apar/IJ05201 for AIX 7.2 which must have been applied to our AIX 7.2 instances -- I don't know if there is an equivalent for AIX 7.1 or the other IBM platforms (@libuv/aix @libuv/ibmi @libuv/zos ). This also explains why the AIX bit of `libuv-in-node` CI in https://github.com/libuv/libuv/pull/3257#issuecomment-1054600676 passed as that ran on AIX 7.2.

IgorTodorovskiIBM commented:
> The situation on AIX is more subtle -- the builds on AIX 7.1 are failing, but on 7.2 they are passing. I found https://www.ibm.com/support/pages/apar/IJ05201 for AIX 7.2 which must have been applied to our AIX 7.2 instances -- I don't know if there is an equivalent for AIX 7.1 or the other IBM platforms (@libuv/aix @libuv/ibmi @libuv/zos ). This also explains why the AIX bit of `libuv-in-node` CI in [#3257 (comment)](https://github.com/libuv/libuv/pull/3257#issuecomment-1054600676) passed as that ran on AIX 7.2.

z/OS does not support F_DUPFD_CLOEXEC, so this would break z/OS.

kadler commented:
This is not supported on IBM i either.
